Issue 136 and 156

benoit74 opened the pull request:
Fix #136 
Fix #156

benoit74 commented:
Some questions / remarks:
- some CLI flags have been removed because they do not makes anymore sense
- I did not achieved to find if it is still possible to create a ZIM without index with zimscraperlib
- I did not achieved to find the documentation regarding the standard minimal Python code guideline (black + isort + flake with 88 chars max length if I did it appropriately in iFixit)
- I did a smoke test with few random books, do you have any idea of which book IDs are particularly relevant to test because they represent a corner case?

kelson42 commented:
> * I did not achieved to find the documentation regarding the standard minimal Python code guideline (black + isort + flake with 88 chars max length if I did it appropriately in iFixit)

It’s here https://github.com/openzim/overview/wiki but clearly not complete.

rgaudin reviewed (changes requested):
Thank you for this large contribution!

I've added a lot of inline comments ; Should all be easy fixes.

The code is not working at the moment though for two reasons:
- the language information passed to export_books that gets transformed for ZIM metadata
- links are broken ; pointing to out-of-bound level (tools.js)
- links also include namespace prefixes that should be removed.

Another issue that I believe have nothing to do with this PR is that when I request a list of formats (epub, html), I do have the proper formats in my ZIM but the UI shows a link to PDF as well because the Book supports PDF (in the DB) but… it's not included because I didn't wanted it.
Code seems to ignore this possibility. We'll open a ticket once this is merged.

FYI Here's the command I used:

```sh
python gutenberg2zim -z gut_fr_test.zim -b 18812 -l fr -f epub,html
```

benoit74 commented:
Thank you for all these comments, I will work on it later this week, seems
pretty easy to fix.

benoit74 commented on gutenbergtozim/database.py:
```diff
@@ -2,20 +2,14 @@
 
 # db = SqliteDatabase('gutenberg.db')
-timeout = 1000 * 60 * 5 * 16
+timeout = 60 * 5 * 16
```
No idea either why it is so high, it is the max indeed. So it caused an integer capacity overflow multiplied by 1000 ;-)
Is it ok for you if I set it to 10 secs ? (I do not see any reason to wait for a lock more than 10 secs, we are not supposed to place any long duration locks on the database with current operations)

rgaudin replied:
I can imagine some lower hardware taking time to process DB query but 10s looks reasonable. We know this exists so should that be an issue, we'll have an idea.

benoit74 replied:
Yep, and this is only a busy timeout, i.e. a timeout for queries which do not achieve to progress because it is waiting for a table which is locked (for modification). In other words, even if the busy timeout is set to 10 secs, you can have queries taking much longer.
Cf. https://sqlite.org/c3ref/busy_timeout.html

benoit74 commented:
@rgaudin, I fixed all inline comments except the one regarding the SQL timeout, see proposition above.

I also fixed following issues (well spotted 👏):
- the language information passed to export_books that gets transformed for ZIM metadata
- links also include namespace prefixes that should be removed

I don't achieve to reproduce this issue (could you provide more info?):
- links are broken ; pointing to out-of-bound level (tools.js)

Regarding the link to the PDF which should not be displayed, I confirm this seems not linked to current PR. Issue is present in many places: cover article (gutenbergtozim/templates/cover_article.html), HTML version of a book (gutenbergtozim/templates/book_infobox.html) and bookshelves (2 times in gutenbergtozim/templates/js/tools.js).

rgaudin commented on gutenberg2zim:
```diff
@@ -51,27 +51,27 @@ help = (
 -b --books=<ids>                Execute the processes for specific books, """
     """separated by commas, or dashes for intervals
--c --concurrency=<nb>           Number of concurrent process for download """
+-p --concurrency=<nb>           Number of concurrent process for download """
```
this has not been answered

benoit74 replied:
I just changed the documentation because it was wrong on this line, the correct parameter is `-p` (cf line 25)

benoit74 replied:
I would have preferred to use `-c` since there is no conflict on this parameter + it is more straightforward, but it would be a breaking change.

rgaudin replied:
Ok let's do it. I'll adapt zimfarm if necessary

rgaudin reviewed (approved):
Sorry for the delay ; have been busy.

This looks good ; There's a couple minor stuff you could look at but this is finally ready. thank you.

kelson42 commented:
@benoit74 Any chance to fix the CodeFactor issues? Usually the only one which are tolerated are the "complex method" ones (which would need a massive revamping and often are older than the PR itself).

rgaudin commented:
> @benoit74 Any chance to fix the CodeFactor issues? Usually the only one which are tolerated are the "complex method" ones (which would need a massive revamping and often are older than the PR itself).

@kelson42 the codefactor issues are from the legacy code. They show up now because the file was touched (to fix the links) but were not introduced now.

benoit74 commented on gutenbergtozim/export.py:
```diff
@@ -151,15 +166,15 @@ def export_skeleton(
+    tpl_path = "Home.html"
+    template = jinja_env.get_template(tpl_path)
+    rendered = template.render(**context)
+    Global.add_item_for(
```
I'm ok with this as well.

benoit74 replied:
Oh no, I'm not, sorry, tpl_path is used twice, that's the reason why I did it like this (once to get the appropriate template, another one to add it to the Zim)

benoit74 commented on gutenbergtozim/shared.py:
```diff
@@ -38,7 +38,7 @@ def setup(filename, language, title, description, name):
             filename=filename,
             main_path="Home.html",
-            favicon_path="favicon.png",
+            favicon_path="illustration",
```
Done

benoit74 commented:
> > @benoit74 Any chance to fix the CodeFactor issues? Usually the only one which are tolerated are the "complex method" ones (which would need a massive revamping and often are older than the PR itself).
> 
> @kelson42 the codefactor issues are from the legacy code. They show up now because the file was touched (to fix the links) but were not introduced now.

Yep, I opened [an issue](https://github.com/openzim/gutenberg/issues/158) to fix this, I would clearly prefer to have this PR merged first. This PR has changed many parts of the code, so delaying it too much is taking a risk of merge conflicts later. And I'm clearly not an expert in JS code, and this tools.js script seems a bit scary to fix without spending some time to understand all its purpose and confirm it is still working as expected afterwards.

benoit74 commented:
> Good to merge ;I'll do it tomorrow; thanks

I can do it if you prefer, @kelson42 granted me sufficient rights. Do you mind if I squash all commits? Are you ok with fast-forward merges? (these are my preferences, but I clearly don't mind to follow other conventions)

rgaudin commented:
Please do. Our convention is to use a merge commit though
